File: Python/boj/Silver/1431.py
import sys

if __name__ == "__main__":
    try:
        sys.stdin = open("input.txt", "r")
    except FileNotFoundError:
        pass

    N = int(input())
    numbers = [input() for _ in range(N)]

    def calNumberSum(number):
        sum = 0

        for i in number:
            if i.isdigit():
                sum += int(i)

        return sum

    numbers.sort(key=lambda x: (len(x), calNumberSum(x), x))
    print("\n".join(numbers))

File: JavaScript/boj/Silver/1431.js
const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const [N, ...input] = fs.readFileSync(filePath).toString().trim().split('\n');

const numbers = input.map((number) => number);

const calNumberSum = (number) => {
  let sum = 0;

  for (let i = 0; i < number.length; i++) {
    const num = Number(number[i]);

    if (!isNaN(num)) sum += num;
  }

  return sum;
};

numbers.sort((a, b) => {
  if (a.length !== b.length) return a.length - b.length;

  if (calNumberSum(a) !== calNumberSum(b)) {
    return calNumberSum(a) - calNumberSum(b);
  }

  return a.localeCompare(b);
});

console.log(numbers.join('\n'));
